packages/contracts-bedrock/test/properties/medusa/ProtocolAtomic.t.sol

```diff
@@ -0,0 +1,180 @@
+        fromIndex = bound(fromIndex, 0, allSuperTokens.length - 1);
+        OptimismSuperchainERC20 sourceToken = OptimismSuperchainERC20(allSuperTokens[fromIndex]);
+        OptimismSuperchainERC20 destinationToken =
+            MESSENGER.crossChainMessageReceiver(address(sourceToken), destinationChainId);
+        // TODO: when implementing non-atomic bridging, allow for the token to
+        // not yet be deployed and funds be recovered afterwards.
+        require(address(destinationToken) != address(0));
+        uint256 sourceBalanceBefore = sourceToken.balanceOf(msg.sender);
+        uint256 sourceSupplyBefore = sourceToken.totalSupply();
+        uint256 destinationBalanceBefore = destinationToken.balanceOf(msg.sender);
+        uint256 destinationSupplyBefore = destinationToken.totalSupply();
+
+        vm.prank(msg.sender);
+        try sourceToken.sendERC20(msg.sender, amount, destinationChainId) {
+            uint256 sourceBalanceAfter = sourceToken.balanceOf(msg.sender);
+            uint256 destinationBalanceAfter = destinationToken.balanceOf(msg.sender);
+            // no free mint
+            assert(sourceBalanceBefore + destinationBalanceBefore == sourceBalanceAfter + destinationBalanceAfter);
+            // 22
+            assert(sourceBalanceBefore - amount == sourceBalanceAfter);
+            assert(destinationBalanceBefore + amount == destinationBalanceAfter);
+            uint256 sourceSupplyAfter = sourceToken.totalSupply();
+            uint256 destinationSupplyAfter = destinationToken.totalSupply();
+            // 23
+            assert(sourceSupplyBefore - amount == sourceSupplyAfter);
+            assert(destinationSupplyBefore + amount == destinationSupplyAfter);
+        } catch {
+            assert(address(destinationToken) == address(sourceToken) || sourceBalanceBefore < amount);
+        }
+    }
+
+    function fuzz_MintSupertoken(uint256 index, uint96 amount) external {
+        index = bound(index, 0, allSuperTokens.length - 1);
+        address addr = allSuperTokens[index];
+        vm.prank(BRIDGE);
+        // medusa calls with different senders by default
+        OptimismSuperchainERC20(addr).mint(msg.sender, amount);
+        // currentValue will be zero if key is not present
+        (,uint256 currentValue) = ghost_totalSupplyAcrossChains.tryGet(MESSENGER.superTokenInitDeploySalts(addr));
+        ghost_totalSupplyAcrossChains.set(MESSENGER.superTokenInitDeploySalts(addr), currentValue + amount);
+    }
+
+    // TODO: will need rework after
+    //   - non-atomic bridge
+    //   - `convert`
+    /// @custom:property-id 24
+    /// @custom:property sum of supertoken total supply across all chains is always equal to convert(legacy, super)-
+    /// convert(super, legacy)
+    /// @dev deliberately not a view method so medusa runs it but not the view methods defined by Test
+    function property_totalSupplyAcrossChainsEqualsMints() external {
```

Comment:
Maybe this?
```suggestion
    function fuzz_totalSupplyAcrossChainsEqualsMints() external {
```

Reply:
this is the only method so far where we manage to _only_ do invariant checking :D

I believe it's worthwhile to give them different names, so later it's more clear if it makes sense to have separate contracts for state transitions or assertions, similar to horsefacts/WETH9